pubsub/iterator.go

```diff
@@ -662,3 +663,50 @@ func maxDuration(x, y time.Duration) time.Duration {
 	}
 	return y
 }
+
+// processResults processes AckResults by referring to errorStatus and errorsMap.
+// The errors returned by the server in `errorStatus` or in `errorsMap`
+// are used to complete the AckResults in `ackResMap` (with a success
+// or error) or to return requests for further retries.
+// Logic is derived from python-pubsub: https://github.com/googleapis/python-pubsub/blob/main/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L161-L220
+func processResults(errorStatus *status.Status, ackResMap map[string]*AckResult, errorsMap map[string]string) ([]*AckResult, []*AckResult) {
```

Comment:
I changed the map's name and value here to your suggestion to avoid the extra string conversion.

I think `errorStatus` should remain as is. In order to initialize `errorsByAckID`, I will need to convert the original grpc error to a `status.Status` (or `APIError`), and passing in an `error` here would mean another re-cast.

Reply:
Hm, actually I was wrong. The map should be `map[string]string` since that's what `APIError.Metadata` returns.